devicetypes/smartthings/zwave-door-window-sensor.src/zwave-door-window-sensor.groovy

```diff
@@ -187,24 +172,26 @@ def zwaveEvent(physicalgraph.zwave.commands.notificationv3.NotificationReport cm
 
 def zwaveEvent(physicalgraph.zwave.commands.wakeupv1.WakeUpNotification cmd) {
 	def event = createEvent(descriptionText: "${device.displayName} woke up", isStateChange: false)
-	def cmds = []
+	def request = []
 	if (!state.MSR) {
-		cmds << command(zwave.manufacturerSpecificV2.manufacturerSpecificGet())
-		cmds << "delay 1200"
+		request << zwave.manufacturerSpecificV2.manufacturerSpecificGet()
 	}
 
-	if (device.currentValue("contact") == null) { // Incase our initial request didn't make it
-		cmds << command(zwave.sensorBinaryV2.sensorBinaryGet(sensorType: zwave.sensorBinaryV2.SENSOR_TYPE_DOOR_WINDOW))
+	if (device.currentValue("contact") == null) {
+		// In case our initial request didn't make it, initial state check no. 3
+		request << zwave.sensorBinaryV2.sensorBinaryGet(sensorType: zwave.sensorBinaryV2.SENSOR_TYPE_DOOR_WINDOW)
 	}
 
 	if (!state.lastbat || now() - state.lastbat > 53 * 60 * 60 * 1000) {
-		cmds << command(zwave.batteryV1.batteryGet())
-	} else {
-		// If we check the battery state we will send NoMoreInfo in the handler for BatteryReport so that we definitely get the report
-		cmds << zwave.wakeUpV1.wakeUpNoMoreInformation().format()
+		request << zwave.batteryV1.batteryGet()
+	}
+
+	def response = response(command(zwave.wakeUpV1.wakeUpNoMoreInformation()))
+	if (request.size() > 0) {
+		response = new physicalgraph.device.HubMultiAction(commands(request, 1000) + ["delay 2000", command(zwave.wakeUpV1.wakeUpNoMoreInformation())])
```

Comment:
For some reason, using request() instead of new HubMultiAction causes errors like:
groovy.lang.MissingMethodException: No signature of method: physicalgraph.device.HubAction.call() is applicable for argument types: (java.util.ArrayList) values: [[9881007204, delay 2000, 9881008408]]
Possible solutions: wait(), any(), wait(long), any(groovy.lang.Closure), each(groovy.lang.Closure), collect()

Probably some problems with dynamic typing.

Reply:
Are you saying that:

`response( commands(request,1000) + ["delay 2000", command(zwave.wakeUpV1.wakeUpNoMoreInformation())] )`

was throwing that error?

I think it's because you took out the `command()` calls which outputs a string. Another subtlety in the original code is that `noMoreInformation` should always be unencapsulated. And your code relies on the device to respond with the battery status within two seconds or else it might have unexpected behavior with the noMoreInformation.

Reply:
response supports the following:

response(String action)
response(physicalgraph.zwave.Command cmd)
response(List actions)

Anyway, command will encapsulate any command if possible. If it is something that you don't want encapsulated, then you wouldn't wrap it with command (or commands), but just append .format(). A good example of the mixed use would be updated() and refresh() in zwave-lock.src/zwave-lock.groovy (replacing secure with command for use here). Then you should be able to take that array or cmds and pass it to response() and then just return that.

So maybe:

`def zwaveEvent(physicalgraph.zwave.commands.wakeupv1.WakeUpNotification cmd) {
	def event = createEvent(descriptionText: "${device.displayName} woke up", isStateChange: false)
	def cmds = []
	if (!state.MSR) {
		cmds << zwave.manufacturerSpecificV2.manufacturerSpecificGet()
	}

	if (device.currentValue("contact") == null) {
		// In case our initial request didn't make it, initial state check no. 3
		cmds << zwave.sensorBinaryV2.sensorBinaryGet(sensorType: zwave.sensorBinaryV2.SENSOR_TYPE_DOOR_WINDOW)
	}

	if (!state.lastbat || now() - state.lastbat > 53 * 60 * 60 * 1000) {
		cmds << zwave.batteryV1.batteryGet()
	}

	def request = []
	if (cmds.size() > 0) {
		request = commands(cmds, 1000)
		request << "delay 2000"
	}
	request << zwave.wakeUpV1.wakeUpNoMoreInformation().format()

	[event, response(request)]
}`

Reply:
@greens 
> Are you saying that:
>
> response( commands(request,1000) + ["delay 2000", command(zwave.wakeUpV1.wakeUpNoMoreInformation())] )
>
> was throwing that error?

Yes.

> I think it's because you took out the command() calls which outputs a string

They are still there. In commands(...) method.

> private commands(commands, delay = 200) {
> 	delayBetween(commands.collect { command(it) }, delay)
> }

> noMoreInformation should always be unencapsulated.

During development this device did not turn the red led off when it was not encapsulated. Manual states it will turn off 2 seconds after receiving noMoreInformation. I will test it without encapsulation 10-20 times more and put it behind mfr/prod/model check or remove it based on results.

> And your code relies on the device to respond with the battery status within two seconds or else it
> might have unexpected behaviour with the noMoreInformation.

I do not think it is a good idea to send it only after we get battery report. Given the amount of "no battery/state after pairing" issues, battery report seems far from guaranteed. And we should always put the device to sleep. Should I make it 20000ms like some other devices?

Reply:
You guys are right, non-encrypted noMoreInformation works well.

@dkirker Thanks for help, your code does not throw any errors.

Reply:
Reasonable point about the `nomoreinformation` after battery report. We'll see how it goes.